build/ci/pr/Jenkinsfile

```diff
@@ -20,6 +20,24 @@ pipeline {
     }
 
     stages {
+        stage('Checkout from GitHub') {
+            steps {
+                checkout scm
+            }
+        }
+        stage('Check if Docker image needs rebuilding') {
+            when {
+                expression {
+                    ifImageNeedsRebuilding()
+                }
+            }
+            environment {
+                IMAGE = "push.docker.elastic.co/eck/eck-ci:latest"
```

Comment:
I don't understand, maybe I'm missing something.
If we run 2 PRs in parallel, and both update go dep, we end up with both replacing our `latest` Docker image in a random order?

Reply:
Maybe we should build a hash of Dockerfile + dep files and use that as the Docker tag?

Reply:
I agree. Using `latest` tags can lead to very subtle issues. However, I do understand that it is tricky to refer to a specific tag in a dynamic environment such as this. 

We have a lot of rarely changing things (gcloud, golangci-lint etc.) and some frequently changing things (dependencies). What if we publish a base image containing the rarely changing things into a registry and explicitly refer to it in the tests? Any changes to the base image will require updating the test config but I don't think it will be too bad. 

Since dependencies could change from test run to test run, it might be easier to download them to disk at start and mount that directory to the containers in subsequent steps (https://jenkins.io/doc/book/pipeline/docker/#caching-data-for-containers).

WDYT?

Reply:
With the hash approach, we could do something like:

```
⟩ md5sum operators/Gopkg.lock operators/Dockerfile local-volume/Gopkg.lock | md5sum | awk '{print $1}'
3494cdfb40865c5cdbdf30fba64439e4
```

Compute this value at every PR (cheap operation), then try to docker pull `docker.elastic.co/eck/eck-3494cdfb40865c5cdbdf30fba64439e4`. If not available, build & push it first.

In most situations, the image would already exist and be reused.

Reply:
Yeah, I planned to do with similar approach after your initial comment. Thanks for detailed info!

@charith-elastic approach most probably will not work in our case. I never used it before, but I suspect that it will keep cached data locally on a Jenkins node. This is not an option for us. Because we create a new node for every build.

Reply:
Fixed!
Now we are using hash as Docker image tag